File: backend/app/services/chess_com_client.py
"""A deliberately serial client for Chess.com's official Published-Data API.

Chess.com's rules (confirmed against their PubAPI FAQ and ToS forum post
before this was written): only the official read-only `api.chess.com/pub/...`
endpoints are used -- never the HTML site, which their ToS explicitly bans
scraping. Serial requests are unlimited; bursts of parallel requests can
trigger 429s. This client enforces a minimum spacing between requests itself
so nothing built on top of it can ever accidentally issue them in parallel --
it's a synchronous `requests.Session`, not an async client, by construction.
"""
import logging
import time
from typing import Any

import requests

logger = logging.getLogger(__name__)

CHESS_COM_API_BASE = "https://api.chess.com/pub"


class ChessComClient:
    def __init__(
        self,
        contact: str,
        min_interval_seconds: float = 1.0,
        max_retries: int = 5,
        timeout_seconds: float = 30.0,
    ):
        self.contact = contact
        self.min_interval_seconds = min_interval_seconds
        self.max_retries = max_retries
        self.timeout_seconds = timeout_seconds
        self._last_request_at: float = 0.0
        self._session = requests.Session()
        self._session.headers["User-Agent"] = f"Chegga/0.1 (personal chess analysis tool; contact: {contact})"

    def _throttle(self) -> None:
        elapsed = time.monotonic() - self._last_request_at
        wait = self.min_interval_seconds - elapsed
        if wait > 0:
            time.sleep(wait)

    def _get(self, url: str) -> dict[str, Any]:
        for attempt in range(1, self.max_retries + 1):
            self._throttle()
            response = self._session.get(url, timeout=self.timeout_seconds)
            self._last_request_at = time.monotonic()

            if response.status_code == 200:
                return response.json()

            if response.status_code == 429 or response.status_code >= 500:
                retry_after = response.headers.get("Retry-After")
                delay = float(retry_after) if retry_after else min(2**attempt, 60)
                logger.warning(
                    "Chess.com API %s on %s (attempt %d/%d) -- backing off %.1fs",
                    response.status_code,
                    url,
                    attempt,
                    self.max_retries,
                    delay,
                )
                time.sleep(delay)
                continue

            response.raise_for_status()

        raise RuntimeError(f"Chess.com API request failed after {self.max_retries} retries: {url}")

    def get_archive_urls(self, username: str) -> list[str]:
        data = self._get(f"{CHESS_COM_API_BASE}/player/{username}/games/archives")
        return data.get("archives", [])

    def get_archive(self, archive_url: str) -> list[dict[str, Any]]:
        data = self._get(archive_url)
        return data.get("games", [])

File: web/src/chessComClient.ts
// Chegga Web — Chess.com Published-Data API client (Phase 1)
//
// Ported from Chegga's own `app/services/chess_com_client.py`, same
// contract: only the official read-only `api.chess.com/pub/...`
// endpoints, serial requests only (never parallel — that's what
// triggers 429s), a minimum spacing enforced client-side, exponential
// backoff honoring `Retry-After` on 429/5xx. A browser hitting the same
// public GET endpoint is exactly what Chess.com's PubAPI is meant for —
// no proxy, no server-side key.

const CHESS_COM_API_BASE = "https://api.chess.com/pub";

export interface ChessComPlayerSide {
  username: string;
  rating?: number;
  result: string;
}

export interface ChessComRawGame {
  uuid: string;
  url?: string;
  pgn?: string;
  time_control?: string;
  time_class?: string;
  rules?: string;
  rated?: boolean;
  end_time?: number;
  eco?: string;
  white: ChessComPlayerSide;
  black: ChessComPlayerSide;
}

/** Thrown for a non-retryable HTTP error so callers can special-case a
 * specific status (e.g. a 404 archive month) without string-matching a
 * generic Error's message. */
export class ChessComHttpError extends Error {
  status: number;

  constructor(status: number, url: string) {
    super(`Chess.com API request failed: ${status} ${url}`);
    this.status = status;
  }
}

export class ChessComClient {
  private minIntervalMs: number;
  private maxRetries: number;
  private lastRequestAt = 0;

  // `contact` no longer has anywhere to go (see the note on `get()` below
  // about why no custom header is sent) — kept as a required param so call
  // sites still read as "who's using this instance," but it's unused.
  constructor(_contact: string, opts: { minIntervalSeconds?: number; maxRetries?: number } = {}) {
    this.minIntervalMs = (opts.minIntervalSeconds ?? 1.0) * 1000;
    this.maxRetries = opts.maxRetries ?? 5;
  }

  private async throttle(): Promise<void> {
    const elapsed = performance.now() - this.lastRequestAt;
    const wait = this.minIntervalMs - elapsed;
    if (wait > 0) await sleep(wait);
  }

  private async get<T>(url: string): Promise<T> {
    for (let attempt = 1; attempt <= this.maxRetries; attempt++) {
      await this.throttle();
      // No custom headers: Chess.com's CORS preflight only allows the
      // browser's own simple-request headers, and a custom header here
      // (an attempted User-Agent/contact-string equivalent, like the
      // backend client sends) makes the preflight fail outright. Browsers
      // also restrict/override User-Agent on fetch regardless, so there's
      // no real equivalent to Python's requests-level UA string from here.
      const response = await fetch(url);
      this.lastRequestAt = performance.now();

      if (response.ok) return (await response.json()) as T;

      if (response.status === 429 || response.status >= 500) {
        const retryAfter = response.headers.get("Retry-After");
        const delayMs = retryAfter ? parseFloat(retryAfter) * 1000 : Math.min(2 ** attempt, 60) * 1000;
        console.warn(
          `Chess.com API ${response.status} on ${url} (attempt ${attempt}/${this.maxRetries}) — backing off ${delayMs}ms`,
        );
        await sleep(delayMs);
        continue;
      }

      throw new ChessComHttpError(response.status, url);
    }

    throw new Error(`Chess.com API request failed after ${this.maxRetries} retries: ${url}`);
  }

  async getArchiveUrls(username: string): Promise<string[]> {
    const data = await this.get<{ archives?: string[] }>(
      `${CHESS_COM_API_BASE}/player/${encodeURIComponent(username)}/games/archives`,
    );
    return data.archives ?? [];
  }

  async getArchive(archiveUrl: string): Promise<ChessComRawGame[]> {
    const data = await this.get<{ games?: ChessComRawGame[] }>(archiveUrl);
    return data.games ?? [];
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
